README.md
[![codecov](https://codecov.io/gh/teranetsrl/oauth2_client/branch/master/graph/badge.svg)](https://codecov.io/gh/teranetsrl/oauth2_client)

# oauth2_client
Simple Flutter library for interacting with OAuth2 servers. It provides convenience classes for interacting with the "usual suspects" (Google, Facebook, LinkedIn, GitHub), but it's particularly suited for implementing clients for custom OAuth2 servers.

The library handles **Authorization Code**, **Client Credentials** and **Implicit Grant** flows.

# Prerequisites #

## Android ##
On Android you must first set the *minSdkVersion* in the *build.gradle* file:
```
defaultConfig {
   ...
   minSdkVersion 18
   ...
```

If at all possible, when registering your application on the OAuth provider **try not to use HTTPS** as the scheme part of the redirect uri, because in that case your application won't intercept the server redirection, as it will be automatically handled by the system browser (at least on Android). Just use a custom scheme, such as "my.test.app" or any other scheme you want.

If the OAuth2 server **allows only HTTPS** uri schemes, refer to the [FAQ](#faq) section.

Again on Android, if your application uses the Authorization Code flow, you first need to modify the *AndroidManifest.xml* file adding the activity `com.linusu.flutter_web_auth.CallbackActivity` with the intent filter needed to open the browser window for the authorization workflow.
The library relies on the flutter_web_auth package to allow the Authorization Code flow.

AndroidManifest.xml

```xml
<activity android:name="com.linusu.flutter_web_auth.CallbackActivity" android:exported="true">
	<intent-filter android:label="flutter_web_auth">
		<action android:name="android.intent.action.VIEW" />
		<category android:name="android.intent.category.DEFAULT" />
		<category android:name="android.intent.category.BROWSABLE" />
		<data android:scheme="my.test.app" />
	</intent-filter>
</activity>
```

## iOS ##
On iOS you need to set the *platform* in the *ios/Podfile* file:
```
platform :ios, '11.0'
```

## Web ##
Web support has been added in the 2.2.0 version, and should be considered preliminary.

On the web platform you **must** register your application using an **HTTPS** redirect uri.

When the authorization code flow is used, the authorization phase will be carried out by opening a popup window to the provider login page.

After the user grants access to your application, the server will redirect the browser to the redirect uri. This page should contain some javascript code to read the _code_ parameter sent by the authorization server and pass it to the parent window through postMessage.

Something like:

```javascript
window.onload = function() {
	const urlParams = new URLSearchParams(window.location.search);
	const code = urlParams.get('code');
	if(code) {
		window.opener.postMessage(window.location.href, _url_of_the_opener_window_);
	}
}
```

**Please note** that the browser can't *securely* store confidential information! The OAuth2Helper class, when used on the web, stores the tokens in the localStorage, and this means they won't be encrypted!

# Installation #

Add the library to your *pubspec.yaml* file:

```yaml
dependencies:
	oauth2_client: ^2.3.3
```

# Usage with the helper class #
The simplest way to use the library is through the *OAuth2Helper* class.
This class transparently handles tokens request/refresh, as well as storing and caching them.

Besides, it implements convenience  methods to transparently perform http requests adding the generated access tokens.

First, instantiate and setup the helper:


```dart
import 'package:oauth2_client_celbux/oauth2_helper.dart';
//We are going to use the google client for this example...
import 'package:oauth2_client_celbux/google_oauth2_client.dart';
import 'package:http/http.dart' as http;

//Instantiate an OAuth2Client...
GoogleOAuth2Client client = GoogleOAuth2Client(
	customUriScheme: 'my.test.app' //Must correspond to the AndroidManifest's "android:scheme" attribute
	redirectUri: 'my.test.app:/oauth2redirect', //Can be any URI, but the scheme part must correspond to the customeUriScheme
);

//Then, instantiate the helper passing the previously instantiated client
OAuth2Helper oauth2Helper = OAuth2Helper(client,
	grantType: OAuth2Helper.AUTHORIZATION_CODE,
	clientId: 'your_client_id',
	clientSecret: 'your_client_secret',
	scopes: ['https://www.googleapis.com/auth/drive.readonly']);

```
In the example we used the Google client, but you can use any other provided client or implement your own (see below).

_Note that the redirect uri has just one slash. This is [required per  Google specs](https://developers.google.com/identity/protocols/oauth2/native-app#step-2:-send-a-request-to-googles-oauth-2.0-server). Other providers could require double slash!_

Now you can use the helper class to perform HTTP requests to the server.

```dart
http.Response resp = helper.get('https://www.googleapis.com/drive/v3/files');
```

The helper will:
 - **check** if a **token** already exists in the **secure storage**
 - if it doesn't exist:
 	- **request the token** using the flow and the parameters specified in the *setAuthorizationParams* call. For example, for the Authorization Code flow this involves opening a web browser for the authorization code and then requesting the actual access token. The token is then **stored in secure storage**.
 - if the token already exists, but is **expired**, a new one is **automatically generated** using the **refresh_token** flow. The token is then stored in secure storage.
 - **Perform** the actual http **request** with the access **token included**.

# Usage without the helper class #
You can use the library without the helper class, using one of the base client classes.

This way tokens won't be automatically stored, nor will be automatically refreshed. Furthermore, you will have to add the access token to http requests by yourself.

```dart
//Import the client you need (see later for available clients)...
import 'myclient.dart'; //Not an actual client!
import 'package:oauth2_client_celbux/access_token_response.dart';

...

//Instantiate the client
client = MyClient(...);

//Request a token using the Authorization Code flow...
AccessTokenResponse tknResp = await client.getTokenWithAuthCodeFlow(
	clientId: 'your_client_id',
	scopes: ['scope1', 'scope2', ...]
);

//Request a token using the Client Credentials flow...
AccessTokenResponse tknResp = await client.getTokenWithClientCredentialsFlow(
	clientId: 'XXX', //Your client id
	clientSecret: 'XXX', //Your client secret
	scopes: ['scope1', 'scope2', ...] //Optional
);

//Or, if you already have a token, check if it is expired and in case refresh it...
if(tknResp.isExpired()) {
	tknResp = client.refreshToken(tknResp.refreshToken);
}
```

## Acessing custom/non standard response fields ##
You can access non standard fields in the response by calling the ```getRespField``` method.

For example:
```dart
AccessTokenResponse tknResp = await client.getTokenWithAuthCodeFlow(
	clientId: 'your_client_id',
	scopes: ['scope1', 'scope2', ...]
);

if(tknResp.isExpired()) {
	var myCustomFieldVal = tknResp.getRespField('my_custom_field');
}
```

# Predefined clients #
The library implements clients for the following services/organizations:

 - Google
 - Facebook
 - LinkedIn
 - GitHub
 - Shopify
 - Spotify

## Google client ##

In order to use this client you need to first configure OAuth2 credentials in the Google API console (https://console.developers.google.com/apis/).

First you need to create a new Project if it doesn't already exists, then you need to create the OAuth2 credentials ("OAuth Client ID").

Select **iOS** as *Application Type*, specify a name for the client and in the *Bundle ID* field insert your custom uri scheme
(for example 'my.test.app', but you can use whatever uri scheme you want).

Then in your code:

```dart
import 'package:oauth2_client_celbux/google_oauth2_client.dart';

OAuth2Client googleClient = GoogleOAuth2Client(
	redirectUri: 'my.test.app:/oauth2redirect', //Just one slash, required by Google specs
	customUriScheme: 'my.test.app'
);
```

Then you can instantiate an helper class or directly use the client methods to acquire access tokens.

## Facebook client ##

In order to use this client you need to first configure OAuth2 credentials in the Facebook dashboard.

Then in your code:

```dart
import 'package:oauth2_client_celbux/facebook_oauth2_client.dart';

OAuth2Client fbClient = FacebookOAuth2Client(
	redirectUri: 'my.test.app://oauth2redirect',
	customUriScheme: 'my.test.app'
);
```

Then you can instantiate an helper class or directly use the client methods to acquire access tokens.

## LinkedIn client ##

In order to use this client you need to first configure OAuth2 credentials. See https://docs.microsoft.com/it-it/linkedin/shared/authentication/authorization-code-flow.

Then in your code:

```dart
import 'package:oauth2_client_celbux/linkedin_oauth2_client.dart';

OAuth2Client liClient = LinkedInOAuth2Client(
	redirectUri: 'my.test.app://oauth2redirect',
	customUriScheme: 'my.test.app'
);
```

Then you can instantiate an helper class or directly use the client methods to acquire access tokens.

## GitHub client ##

In order to use this client you need to first create a new OAuth2 App in the GittHub Developer Settings (https://github.com/settings/developers)

Then in your code:

```dart
import 'package:oauth2_client_celbux/github_oauth2_client.dart';

OAuth2Client ghClient = GitHubOAuth2Client(
	redirectUri: 'my.test.app://oauth2redirect',
	customUriScheme: 'my.test.app'
);
```

Then you can instantiate an helper class or directly use the client methods to acquire access tokens.

# Implementing your own client #
Implementing your own client is quite simple, and often it requires only few lines of code.

In the majority of cases you only need to extend the base *OAuth2Client* class and configure the proper endpoints for the authorization and token url.

```dart
import 'package:oauth2_client_celbux/oauth2_client.dart';

class MyOAuth2Client extends OAuth2Client {
  MyOAuth2Client({required String redirectUri, required String customUriScheme}): super(
    authorizeUrl: 'https://...', //Your service's authorization url
    tokenUrl: 'https://...', //Your service access token url
    redirectUri: redirectUri,
    customUriScheme: customUriScheme
  );
}
```

## Open ID support ##
Open ID Connect is currently in development. Stay tuned for updates!

## <a name="faqe"></a>FAQ ##

### Tokens are not getting stored! ###
If when using the helper class the tokens seem to not getting stored, it could be that the requested scopes differs from those returned by the Authorization server.

OAuth2 specs state that the server could optionally return the granted scopes. The OAuth2Helper, when storing an access token, keeps track of the scopes it has been granted for, so the next time a token is needed for one or more of those scopes, it will be readily available without performing another authorization flow.

If the client requests an authorization grant for scopes "A" and "B", but the server for some reason returns a token valid for scope "A" only, that token will be stored along with scope "A", and not "B".
This means that the next time the client will need a token for scopes "A" and "B", the helper will check its storage looking for a token for both "A" and "B", but will only find a token valid for "A", so it will start a new authorization process.

To verify that the requested scopes are really the ones granted on the server, you can use something like the following:

```dart
var client = OAuth2Client(
  authorizeUrl: <YOUR_AUTHORIZE_URL>,
  tokenUrl: <YOUR_TOKEN_URL>,
  redirectUri: <YOUR_REDIRECT_URI>,
  customUriScheme: <YOUR_CUSTOM_SCHEME>);

var tknResp = await client.getTokenWithAuthCodeFlow(
  clientId: <YOUR_CLIENT_ID>,
  scopes: [
	  <LIST_OF_SCOPES>
  ]);

print(tknResp.httpStatusCode);
print(tknResp.error);
print(tknResp.expirationDate);
print(tknResp.scope);
```

Apart from the order, the printed scopes should correspond **exactly** to the ones you requested.

### I get an error *PlatformException(CANCELED, User canceled login, null, null)* on Android ###
Please make sure you modified the *AndroidManifest.xml* file adding the  ```flutter_web_auth.CallbackActivity``` and the intent filter needed to open the browser window for the authorization workflow.

The AndroidManifest.xml file must contain the ```flutter_web_auth.CallbackActivity``` activity. Copy and paste the below code and CHANGE the value of `android:scheme` to match the scheme used in the redirect uri:

```xml
<activity android:name="com.linusu.flutter_web_auth.CallbackActivity" android:exported="true">
	<intent-filter android:label="flutter_web_auth">
		<action android:name="android.intent.action.VIEW" />
		<category android:name="android.intent.category.DEFAULT" />
		<category android:name="android.intent.category.BROWSABLE" />
		<data android:scheme="my.test.app" />
	</intent-filter>
</activity>
```

### Can I use https instead of a custom scheme? ###

If you want to use an HTTPS url as the redirect uri, you must setup it as an [App Link](https://developer.android.com/training/app-links/index.html).
First you need to specify both the ```android:host``` and ```android:pathPrefix``` attributes, as long as the ```android:autoVerify="true"``` attribute in the intent-filter tag inside the _AndroidManifest.xml_:

```xml
<activity android:name="com.linusu.flutter_web_auth.CallbackActivity" >
	<intent-filter android:label="flutter_web_auth" android:autoVerify="true">
		<action android:name="android.intent.action.VIEW" />
		<category android:name="android.intent.category.DEFAULT" />
		<category android:name="android.intent.category.BROWSABLE" />

		<data android:scheme="https"
				android:host="www.myapp.com"
				android:pathPrefix="/oauth2redirect" />
	</intent-filter>
</activity>
```

Then you need to [prove ownership](https://developer.android.com/training/app-links/verify-site-associations) of the domain host by publishing a [Digital Asset Links](https://developers.google.com/digital-asset-links/v1/getting-started) JSON file on your website. This involves generating an [App signing key](https://developer.android.com/studio/publish/app-signing) and signing your app with it.


CHANGELOG.md
## [2.3.5] - 2022/03/25
* Restored client_id in oauth2_client.dart
## [2.3.4] - 2022/03/24
* Updated oauth2_utils.dart to generate url from token and auth urls already containing parameters
## [2.3.3] - 2022/03/14
* Expose the BaseStorage object (thanks [Jon Salmon](https://github.com/Jon-Salmon))
* Improved documentation for android Manifest modification (thanks [supermar1010](https://github.com/supermar1010))
* Update flutter_web_auth dependency
## [2.3.2] - 2022/01/17
* Fix: support fetching a new token when expired without a refresh token (thanks [Tiernan](https://github.com/nvx))
* Updated dependencies (thanks [Menelphor](https://github.com/Menelphor))
* Fix linter warnings

## [2.3.1] - 2021/12/13
* Small fix
## [2.3.0] - 2021/11/03
* Add Spotify client (thanks [mauriciocartagena](https://github.com/mauriciocartagena))
* Allow passing custom WebAuth class to OAuth2Helper (thanks [Jon Salmon](https://github.com/Jon-Salmon))

## [2.2.7] - 2021/10/07
* Fix: null check operator in WebAuth class (thanks [jakub-bacic](https://github.com/jakub-bacic))

## [2.2.6] - 2021/10/03
* Allow passing optional parameters to web_auth "authenticate" method

## [2.2.5] - 2021/09/29
* Fix for authorization url params encoding

## [2.2.4] - 2021/09/28
* Support ephemeral sessions (thanks [ThexXTURBOXx](https://github.com/ThexXTURBOXx))
* Fixes (URI components encoding)

## [2.2.2] - 2021/06/27
* Fixes (scope handling)

## [2.2.1] - 2021/06/23
* Fixes (incorrect scope handling with implicit grant, send empty client secret if specified)
* Added Reddit client (thanks [lavalleeale](https://github.com/lavalleeale))
## [2.2.0] - 2021/05/29
* Web platform support!
## [2.1.0] - 2021/04/15
* AccessTokenResponse refactorization. It is now possible to retrieve custom response fields through the ```getRespField``` method
* Scopes separator can be configured with the OAuth2Client's ```scopeSeparator``` param
## [2.0.1] - 2021/04/06
* Complete migration to sound null safety
* Updated dependencies
## [2.0.0-nullsafety] - 2021/03/30
* Migration to null safety
* Deprecated OAuth2Helper.setAuthorizationParams method
## [1.8.0] - 2021/03/28
* Add compatibility with http 0.13 (thanks [bangfalse](https://github.com/bangfalse))
* Allow passing credentials location in request body or header (thanks [sbu-WBT](https://github.com/wbt-solutions))
* Added Shopify client (thanks [sbu-WBT](https://github.com/wbt-solutions))
## [1.7.1] - 2021/02/25
* Fixes (make httpClient optional again in OAuth2Helper)

## [1.7.0] - 2021/02/24
* Added PUT, PATCH and HEAD methods to the OAuth2Helper class

## [1.6.2] - 2021/01/31
* Allows iOS token reading from storage when invoked in background task
* Small fixes

## [1.6.1] - 2020/10/29
* Changed helper's token retrieval method

## [1.6.0] - 2020/10/13
* Added "delete" request method to helper class
* OAuth requests refactorization
* Fixes (fallback token_type parameter value)

## [1.5.1] - 2020/09/15
* Made "state" parameter optional
* Fixes (refreshToken)

## [1.5.0] - 2020/09/12
* Add implicit grant flow

## [1.4.6] - 2020/09/03
* Allow disabling PKCE when using OAuth2Helper

## [1.4.5] - 2020/08/23
* Bugfixes (check "expires_in" parameter type)

## [1.4.4] - 2020/08/10
* Bugfixes (add null-aware operators)

## [1.4.3] - 2020/07/09
* Token storage fix when scopes are empty

## [1.4.2] - 2020/06/27
* Revocation token fixes

## [1.4.1] - 2020/06/19
* Small fixes

## [1.4.0] - 2020/06/15
* scopes parameter become ooptional (as per the OAuth2 specs)
* custom query parameters sent back from the authorization code response can now be retrieved throught the AuthorizationCode.getQueryParam method
* added _afterAuthorizationCodeCb_ callback to handle use cases for access token requests
* added custom params handling to the Authorization Code and the Access Token Requests

## [1.3.2] - 2020/06/08
* Minor bugfixes.

## [1.3.1] - 2020/06/05
* Handled situations in which no new refresh token is returned upon a refresh flow.

## [1.3.0] - 2020/05/27
* Added revocation token ("logout") process
* Refresh token flow is more spec compliant
* Updated dependencies

## [1.2.4] - 2020/05/10
* Bugfixes (optional scopes handling in the Access Token Response)

## [1.2.3] - 2020/05/06
* Bugfixes (multiple scopes handling)

## [1.2.2] - 2020/05/03
* Added trim for "scope" parameter

## [1.2.1] - 2020/04/29
* Bugfixes

## [1.2.0] - 2020/04/22
* Added the _headers_ parameter to the _oauth2_helper_'s _get_ and _post_ methods.

## [1.1.2] - 2020/04/13
* Updated dependencies

## [1.1.1] - 2020/04/06
* Simplified helper set up

## [1.1.0] - 2020/04/02
* Implemented GitHub client. Added the possibility to specify custom headers to the Access Token request. Partial OAuth2Helper refactorization.

## [1.0.2] - 2020/03/10
* Added example, minor bugfixes.

## [1.0.1] - 2020/03/09
* Bugfixes, added test cases.

## [1.0.0] - 2020/01/05
* First public release.


lib/oauth2_client.dart
import 'dart:convert';

import 'package:http/http.dart' as http;
import 'package:oauth2_client/access_token_response.dart';
import 'package:oauth2_client/authorization_response.dart';
import 'package:oauth2_client/oauth2_response.dart';
import 'package:oauth2_client/src/oauth2_utils.dart';
// import 'package:oauth2_client/src/web_auth.dart';

import 'src/base_web_auth.dart';
import 'src/web_auth.dart'
    // ignore: uri_does_not_exist
    if (dart.library.io) 'src/io_web_auth.dart'
    // ignore: uri_does_not_exist
    if (dart.library.html) 'src/browser_web_auth.dart';

import 'package:random_string/random_string.dart';

enum CredentialsLocation { HEADER, BODY }

/// Base class that implements OAuth2 authorization flows.
///
/// It currently supports the following grants:
/// * Authorization Code
/// * Client Credentials
///
/// For the Authorization Code grant, PKCE is used by default. If you need to disable it, pass the 'enablePKCE' param to false.
///
/// You can use directly this class, but normally you want to extend it and implement your own client.
/// When instantiating the client, pass your custom uri scheme in the [customUriScheme] field.
/// Normally you would use something like <customUriScheme>:/oauth for the [redirectUri] field.
/// For Android only you must add an intent filter in your AndroidManifest.xml file to enable the custom uri handling.
/// <activity android:name="com.linusu.flutter_web_auth.CallbackActivity" >
///   <intent-filter android:label="flutter_web_auth">
///     <action android:name="android.intent.action.VIEW" />
///     <category android:name="android.intent.category.DEFAULT" />
///     <category android:name="android.intent.category.BROWSABLE" />
///     <data android:scheme="com.teranet.app" />
///   </intent-filter>
/// </activity>
class OAuth2Client {
  String redirectUri;
  String customUriScheme;

  String tokenUrl;
  String? refreshUrl;
  String? revokeUrl;
  String authorizeUrl;
  String scopeSeparator;

  Map<String, String> _accessTokenRequestHeaders = {};

  BaseWebAuth webAuthClient = createWebAuth();
  CredentialsLocation credentialsLocation;

  /// Creates a new client instance with the following parameters:
  ///
  /// * [authorizeUrl]: the url that must be used to fetch authorization codes (for Authorization Code flows)
  /// * [tokenUrl]: the url to be used for generating the OAuth2 access tokens
  /// * [refreshUrl]: the url that must be used for refreshing an Access Token
  /// * [revokeUrl]: the url to be invoked for token revocation
  /// * [redirectUri]: the redirect uri defined in the provider's client registration panel
  /// * [customUriScheme]: the scheme used for the redirect uri
  /// * [credentialsLocation]: where the credentials (client ID / client secret) should be passed (header / body)
  /// * [scopeSeparator]: the separator that has to be used to serialize scopes in the token request
  OAuth2Client(
      {required this.authorizeUrl,
      required this.tokenUrl,
      this.refreshUrl,
      this.revokeUrl,
      required this.redirectUri,
      required this.customUriScheme,
      this.credentialsLocation = CredentialsLocation.HEADER,
      this.scopeSeparator = ' '});

  /// Requests an Access Token to the OAuth2 endpoint using the Implicit grant flow (https://tools.ietf.org/html/rfc6749#page-31)
  Future<AccessTokenResponse> getTokenWithImplicitGrantFlow(
      {required String clientId,
      List<String>? scopes,
      bool enableState = true,
      String? state,
      httpClient,
      BaseWebAuth? webAuthClient,
      Map<String, dynamic>? webAuthOpts}) async {
    httpClient ??= http.Client();
    webAuthClient ??= this.webAuthClient;

    if (enableState) state ??= randomAlphaNumeric(25);

    final authorizeUrl = getAuthorizeUrl(
        clientId: clientId,
        responseType: 'token',
        scopes: scopes,
        enableState: enableState,
        state: state,
        redirectUri: redirectUri);

    // Present the dialog to the user
    final result = await webAuthClient.authenticate(
        url: authorizeUrl,
        callbackUrlScheme: customUriScheme,
        redirectUrl: redirectUri,
        opts: webAuthOpts);

    final fragment = Uri.splitQueryString(Uri.parse(result).fragment);

    if (enableState) {
      final checkState = fragment['state'];
      if (state != checkState) {
        throw Exception(
            '"state" parameter in response doesn\'t correspond to the expected value');
      }
    }

    return AccessTokenResponse.fromMap({
      'access_token': fragment['access_token'],
      'token_type': fragment['token_type'],
      'scope': fragment['scope'] ?? scopes,
      'expires_in': fragment['expires_in'],
      'http_status_code': 200
    });
  }

  /// Requests an Access Token to the OAuth2 endpoint using the Authorization Code Flow.
  Future<AccessTokenResponse> getTokenWithAuthCodeFlow(
      {required String clientId,
      List<String>? scopes,
      String? clientSecret,
      bool enablePKCE = true,
      bool enableState = true,
      String? state,
      String? codeVerifier,
      Function? afterAuthorizationCodeCb,
      Map<String, dynamic>? authCodeParams,
      Map<String, dynamic>? accessTokenParams,
      httpClient,
      BaseWebAuth? webAuthClient,
      Map<String, dynamic>? webAuthOpts}) async {
    AccessTokenResponse? tknResp;

    String? codeChallenge;

    if (enablePKCE) {
      codeVerifier ??= randomAlphaNumeric(80);

      codeChallenge = OAuth2Utils.generateCodeChallenge(codeVerifier);
    }

    var authResp = await requestAuthorization(
        webAuthClient: webAuthClient,
        clientId: clientId,
        scopes: scopes,
        codeChallenge: codeChallenge,
        enableState: enableState,
        state: state,
        customParams: authCodeParams,
        webAuthOpts: webAuthOpts);

    if (authResp.isAccessGranted()) {
      if (afterAuthorizationCodeCb != null) afterAuthorizationCodeCb(authResp);

      tknResp = await requestAccessToken(
          httpClient: httpClient,
          //If the authorization request was successfull, the code must be set
          //otherwise an exception is raised in the OAuth2Response constructor
          code: authResp.code!,
          clientId: clientId,
          scopes: scopes,
          clientSecret: clientSecret,
          codeVerifier: codeVerifier,
          customParams: accessTokenParams);
    } else {
      tknResp = AccessTokenResponse.errorResponse();
    }

    return tknResp;
  }

  /// Requests an Access Token to the OAuth2 endpoint using the Client Credentials flow.
  Future<AccessTokenResponse> getTokenWithClientCredentialsFlow(
      {required String clientId,
      required String clientSecret,
      List<String>? scopes,
      httpClient}) async {
    var params = <String, String>{'grant_type': 'client_credentials'};

    if (scopes != null && scopes.isNotEmpty) {
      params['scope'] = serializeScopes(scopes);
    }

    var response = await _performAuthorizedRequest(
        url: tokenUrl,
        clientId: clientId,
        clientSecret: clientSecret,
        params: params,
        httpClient: httpClient);

    return http2TokenResponse(response, requestedScopes: scopes);
  }

  /// Requests an Authorization Code to be used in the Authorization Code grant.
  Future<AuthorizationResponse> requestAuthorization(
      {required String clientId,
      List<String>? scopes,
      String? codeChallenge,
      bool enableState = true,
      String? state,
      Map<String, dynamic>? customParams,
      BaseWebAuth? webAuthClient,
      Map<String, dynamic>? webAuthOpts}) async {
    webAuthClient ??= this.webAuthClient;

    if (enableState) {
      state ??= randomAlphaNumeric(25);
    }

    final authorizeUrl = getAuthorizeUrl(
        clientId: clientId,
        redirectUri: redirectUri,
        scopes: scopes,
        enableState: enableState,
        state: state,
        codeChallenge: codeChallenge,
        customParams: customParams);

    // Present the dialog to the user
    final result = await webAuthClient.authenticate(
        url: authorizeUrl,
        callbackUrlScheme: customUriScheme,
        redirectUrl: redirectUri,
        opts: webAuthOpts);

    return AuthorizationResponse.fromRedirectUri(result, state);
  }

  /// Requests and Access Token using the provided Authorization [code].
  Future<AccessTokenResponse> requestAccessToken(
      {required String code,
      required String clientId,
      String? clientSecret,
      String? codeVerifier,
      List<String>? scopes,
      Map<String, dynamic>? customParams,
      httpClient}) async {
    final params = getTokenUrlParams(
        code: code,
        redirectUri: redirectUri,
        codeVerifier: codeVerifier,
        clientId: clientId,
        customParams: customParams);

    var response = await _performAuthorizedRequest(
        url: tokenUrl,
        clientId: clientId,
        clientSecret: clientSecret,
        params: params,
        headers: _accessTokenRequestHeaders,
        httpClient: httpClient);

    return http2TokenResponse(response, requestedScopes: scopes);
  }

  /// Refreshes an Access Token issuing a refresh_token grant to the OAuth2 server.
  Future<AccessTokenResponse> refreshToken(String refreshToken,
      {httpClient, required String clientId, String? clientSecret}) async {
    final Map params = getRefreshUrlParams(refreshToken: refreshToken);

    var response = await _performAuthorizedRequest(
        url: _getRefreshUrl(),
        clientId: clientId,
        clientSecret: clientSecret,
        params: params,
        httpClient: httpClient);

    return http2TokenResponse(response);
  }

  /// Revokes both the Access and the Refresh tokens in the provided [tknResp]
  Future<OAuth2Response> revokeToken(AccessTokenResponse tknResp,
      {String? clientId, String? clientSecret, httpClient}) async {
    var tokenRevocationResp = await revokeAccessToken(tknResp,
        clientId: clientId, clientSecret: clientSecret, httpClient: httpClient);
    if (tokenRevocationResp.isValid()) {
      tokenRevocationResp = await revokeRefreshToken(tknResp,
          clientId: clientId,
          clientSecret: clientSecret,
          httpClient: httpClient);
    }

    return tokenRevocationResp;
  }

  /// Revokes the Access Token in the provided [tknResp]
  Future<OAuth2Response> revokeAccessToken(AccessTokenResponse tknResp,
      {String? clientId, String? clientSecret, httpClient}) async {
    return await _revokeTokenByType(tknResp, 'access_token',
        clientId: clientId, clientSecret: clientSecret, httpClient: httpClient);
  }

  /// Revokes the Refresh Token in the provided [tknResp]
  Future<OAuth2Response> revokeRefreshToken(AccessTokenResponse tknResp,
      {String? clientId, String? clientSecret, httpClient}) async {
    return await _revokeTokenByType(tknResp, 'refresh_token',
        clientId: clientId, clientSecret: clientSecret, httpClient: httpClient);
  }

  /// Generates the url to be used for fetching the authorization code.
  String getAuthorizeUrl(
      {required String clientId,
      String responseType = 'code',
      String? redirectUri,
      List<String>? scopes,
      bool enableState = true,
      String? state,
      String? codeChallenge,
      Map<String, dynamic>? customParams}) {
    final params = <String, dynamic>{
      'response_type': responseType,
      'client_id': clientId
    };

    if (redirectUri != null && redirectUri.isNotEmpty) {
      params['redirect_uri'] = redirectUri;
    }

    if (scopes != null && scopes.isNotEmpty) {
      params['scope'] = serializeScopes(scopes);
    }

    if (enableState && state != null && state.isNotEmpty) {
      params['state'] = state;
    }

    if (codeChallenge != null && codeChallenge.isNotEmpty) {
      params['code_challenge'] = codeChallenge;
      params['code_challenge_method'] = 'S256';
    }

    if (customParams != null) {
      params.addAll(customParams);
    }

    return OAuth2Utils.addParamsToUrl(authorizeUrl, params);
  }

  /// Returns the parameters needed for the authorization code request
  Map<String, dynamic> getTokenUrlParams(
      {required String code,
      String? redirectUri,
      String? codeVerifier,
      String? clientId,
      Map<String, dynamic>? customParams}) {
    final params = <String, dynamic>{
      'grant_type': 'authorization_code',
      'code': code
    };

    if (redirectUri != null && redirectUri.isNotEmpty) {
      params['redirect_uri'] = redirectUri;
    }

    if (clientId != null && clientId.isNotEmpty) {
      params['client_id'] = clientId;
    }

/*
    //If a client secret has been specified, it will be sent in the "Authorization" header instead of a body parameter...
    if (clientSecret == null || clientSecret.isEmpty) {
      
    }
*/

    if (codeVerifier != null && codeVerifier.isNotEmpty) {
      params['code_verifier'] = codeVerifier;
    }

    if (customParams != null) {
      params.addAll(customParams);
    }

    return params;
  }

  /// Performs a post request to the specified [url],
  /// adding authentication credentials as described here: https://tools.ietf.org/html/rfc6749#section-2.3
  Future<http.Response> _performAuthorizedRequest(
      {required String url,
      required String clientId,
      String? clientSecret,
      Map? params,
      Map<String, String>? headers,
      httpClient}) async {
    httpClient ??= http.Client();

    headers ??= {};
    params ??= {};

    //If a client secret has been specified, it will be sent in the "Authorization" header instead of a body parameter...
    if (clientSecret == null) {
      if (clientId.isNotEmpty) {
        params['client_id'] = clientId;
      }
    } else {
      switch (credentialsLocation) {
        case CredentialsLocation.HEADER:
          headers.addAll(getAuthorizationHeader(
            clientId: clientId,
            clientSecret: clientSecret,
          ));
          break;
        case CredentialsLocation.BODY:
          params['client_id'] = clientId;
          params['client_secret'] = clientSecret;
          break;
      }
    }

    var response =
        await httpClient.post(Uri.parse(url), body: params, headers: headers);

    return response;
  }

  Map<String, String> getAuthorizationHeader(
      {required String clientId, String? clientSecret}) {
    var headers = <String, String>{};

    if ((clientId.isNotEmpty) && (clientSecret != null)) {
      var credentials =
          base64.encode(utf8.encode(clientId + ':' + clientSecret));

      headers['Authorization'] = 'Basic ' + credentials;
    }

    return headers;
  }

  /// Returns the parameters needed for the refresh token request
  Map<String, String> getRefreshUrlParams({required String refreshToken}) {
    final params = <String, String>{
      'grant_type': 'refresh_token',
      'refresh_token': refreshToken
    };

    return params;
  }

  AccessTokenResponse http2TokenResponse(http.Response response,
      {List<String>? requestedScopes}) {
    return AccessTokenResponse.fromHttpResponse(response,
        requestedScopes: requestedScopes);
  }

  String serializeScopes(List<String> scopes) {
    return scopes.map((s) => s.trim()).join(scopeSeparator);
  }

  /// Revokes the specified token [type] in the [tknResp]
  Future<OAuth2Response> _revokeTokenByType(
      AccessTokenResponse tknResp, String tokenType,
      {String? clientId, String? clientSecret, httpClient}) async {
    var resp = OAuth2Response();

    if (revokeUrl == null) return resp;

    httpClient ??= http.Client();

    var token = tokenType == 'access_token'
        ? tknResp.accessToken
        : tknResp.refreshToken;

    if (token != null) {
      var params = {'token': token, 'token_type_hint': tokenType};

      if (clientId != null) params['client_id'] = clientId;
      if (clientSecret != null) params['client_secret'] = clientSecret;

      http.Response response =
          await httpClient.post(Uri.parse(revokeUrl!), body: params);

      resp = http2TokenResponse(response);
    }

    return resp;
  }

  String _getRefreshUrl() {
    return refreshUrl ?? tokenUrl;
  }

  set accessTokenRequestHeaders(Map<String, String> headers) {
    _accessTokenRequestHeaders = headers;
  }
}
